File: data_structures/hash_tables/hash_table_chains/hash_table_chains.py
# python3
"""
Implement a map for storing strings based on a hash table with lists chaining.
Polynomial hash function for hashing a string:
h(S) = (sum(S[i]*x**i) mod p) mod m
where
S[i] - ASCII code of i-th symbol of S
p = 1000000007
x = 263

Your program should support
 - add string
 - del string
 - find string
 - check i
"""


class Chain:
    def __init__(self, value):
        self.next = None
        self.value = value


class HashTable:
    _multiplier = 263
    _prime = 1000000007

    def __init__(self, cardinality):
        self.arr = [None] * cardinality
        self.cardinality = cardinality

    def add(self, value):
        new_chain = Chain(value)
        i = self._hash(value)
        chain = self.arr[i]

        if chain:
            while chain:
                # if such value exists in hash table - ignore query
                if chain.value == value:
                    return
                else:
                    chain = chain.next

            new_chain.next = self.arr[i]

        self.arr[i] = new_chain

    def find(self, value):
        chain = self.arr[self._hash(value)]
        while chain:
            if chain.value == value:
                return "yes"

            chain = chain.next

        return "no"

    def delete(self, value):
        i = self._hash(value)
        chain = self.arr[i]
        prev_chain = None
        while chain:
            if chain.value == value:
                if prev_chain:
                    prev_chain.next = chain.next
                else:
                    self.arr[i] = chain.next

                return
            else:
                prev_chain = chain
                chain = chain.next

    def check(self, i):
        chain = self.arr[i]
        while chain:
            print(chain.value, end=" ")
            chain = chain.next

        print("")

    def _hash(self, string):
        acc = 0
        for i in range(len(string)):
            acc += ord(string[i]) * self._multiplier ** i

        return acc % self._prime % self.cardinality


def main():
    cardinality = int(input())
    n = int(input())
    ht = HashTable(cardinality)

    for _ in range(n):
        command, value = input().split()
        if command == "add":
            ht.add(value)
        elif command == "del":
            ht.delete(value)
        elif command == "find":
            print(ht.find(value))
        else:
            ht.check(int(value))


if __name__ == "__main__":
    main()

File: data_structures/hash_tables/hash_table_chains/hash_table_chains.go
package main

import (
	"fmt"
)

type HashTable struct {
	cardinality int
	arr         []*Chain
	prime       int
	multiplier  int
}

type Chain struct {
	next  *Chain
	value string
}

func (ht *HashTable) Hash(valuePtr *string) int {
	var acc int
	asciiCodes := []rune(*valuePtr)
	for i := len(asciiCodes) - 1; i >= 0; i-- {
		acc = (acc * ht.multiplier + int(asciiCodes[i])) % ht.prime
	}
	return acc % ht.prime % ht.cardinality
}

func (ht *HashTable) Add(value string) {
	newChain := Chain{value: value}
	i := ht.Hash(&value)
	chainPtr := ht.arr[i]

	if chainPtr != nil {
		for chainPtr != nil {
			if chainPtr.value == value {
				return
			} else {
				chainPtr = chainPtr.next
			}
		}
		newChain.next = ht.arr[i]
	}

	ht.arr[i] = &newChain
}

func (ht *HashTable) Find(value string) string {
	chainPtr := ht.arr[ht.Hash(&value)]
	for chainPtr != nil {
		if chainPtr.value == value {
			return "yes"
		}

		chainPtr = chainPtr.next
	}

	return "no"
}

func (ht *HashTable) Delete(value string) {
	i := ht.Hash(&value)
	chainPtr := ht.arr[i]
	var prevPtr *Chain

	for chainPtr != nil {
		if chainPtr.value == value {
			if prevPtr != nil {
				prevPtr.next = chainPtr.next
			} else {
				ht.arr[i] = chainPtr.next
			}
			return

		} else {
			prevPtr = chainPtr
			chainPtr = chainPtr.next
		}
	}
}

func (ht *HashTable) Check(i int) {
	chainPtr := ht.arr[i]
	res := ""

	for chainPtr != nil {
		res += chainPtr.value
		if chainPtr.next != nil {
			res += " "
		} else {
			break
		}
		chainPtr = chainPtr.next
	}

	fmt.Println(res)
}

func main() {
	var n int
	var m int
	var value, command string

	_, _ = fmt.Scanln(&m)
	_, _ = fmt.Scanln(&n)

	ht := HashTable{cardinality: m, prime: 1000000007, multiplier: 263}
	ht.arr = make([]*Chain, ht.cardinality)

	for i := 0; i < n; i++ {
		_, _ = fmt.Scan(&command)
		if command == "check" {
			var valueInt int
			_, _ = fmt.Scanln(&valueInt)
			ht.Check(valueInt)

		} else {
			_, _ = fmt.Scanln(&value)
			if command == "add" {
				ht.Add(value)

			} else if command == "find" {
				fmt.Println(ht.Find(value))

			} else {
				ht.Delete(value)
			}
		}
	}
}
